File: incre-tests/synduce/expressions/norm_arith.f
Inductive ArithExpr = int Int | plus {ArithExpr, ArithExpr} | minus {ArithExpr, ArithExpr};
Inductive NormExpr = nint Int | nplus {NormExpr, NormExpr};

repr = fix (
  \f: NormExpr -> Compress ArithExpr. \e: NormExpr.
  match e with
    nint i -> int i
  | nplus {e1, e2} -> plus {f e1, f e2}
  end
);

spec = fix (
  \f: ArithExpr -> Int. \e: ArithExpr.
  match e with
    int i -> i
  | plus {e1, e2} -> + (f e1) (f e2)
  | minus {e1, e2} -> - (f e1) (f e2)
  end
);

main = \e: NormExpr. spec (repr e);

File: ml-benchmark/synduce/expressions/norm_arith.ml
type arith_expr =
  | Int of int
  | Plus of arith_expr * arith_expr
  | Minus of arith_expr * arith_expr

type norm_expr =
  | NInt of int
  | NPlus of norm_expr * norm_expr

val repr: norm_expr -> arith_expr compress
let rec repr = function
  | NInt i -> Int i
  | NPlus (a, b) -> Plus (repr a, repr b)

let rec spec = function
  | Int i -> i
  | Plus (a, b) -> spec a + spec b
  | Minus (a, b) -> spec a - spec b

let program e = spec (repr e)